### PR TITLE
add zeppelin.livy.share_existing_session option

### PR DESCRIPTION
**Pull request by ruseel**
If you want to share livy session between multiple zeppelin user, There is no option now.
After this commit, setting zeppelin.livy.share_existing_session to true
make LivyInterpreter use an existing livy session
(first session returned by /sessions)

### What is this PR for?
add zeppelin.livy.share_existing_session option. 
zeppelin.livy.share_existing_session is true, list exsiting livy session and select first session. 
by doing so, we can share livy session between zeppelin livy interpreters. 

### What type of PR is it?
Feature

### How should this be tested?

* setup livy server 
* run zeppelin server
* setup zeppelin livy interpreter(change zeppelin.livy.share_existing_session to true) 
* make new note 
* make another note 
* audit logs/zeppelin-interpreter-*.log

### Questions:
* Does the licenses files need update? No 
* Is there breaking changes for older versions? No
* Does this needs documentation? Yes

### CODE REVIEW

File: livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterpreter.java [hunk 1]
```diff
@@ -119,25 +144,25 @@ public void close() {
     }
   }
 
-  protected void initLivySession() throws LivyException {
```
Comment (ruseel) [inline livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterpreter.java]: initLivySession() was setting `this.sessionInfo`. 

changed name to createSession and removed assignment to`this.sessionInfo` to its caller

**Comment by swkimme** [comment]
@jerryshao may I ask you to review this PR?

**Comment by zjffdu** [comment]
@ruseel @swkimme livy session is shared across notes, What kind of scenarios here do you mean ?

**Comment by ruseel** [comment]
I am trying to use livy as shown in https://youtu.be/gtql7z_kj3Y Zeppelin and livy bring multi tenancy to interactive data analysis 20:02. 

Share livy session between multiple user.

**Comment by ruseel** [comment]
What I want to accomplish eventually is providing a single spark cluster to four people and interactive query they issued  is run under a single pool, with a fair scheduler.

**Comment by ruseel** [comment]
> livy session is shared across notes, What kind of scenarios here do you mean?

I guessed that "Livy session is not shared across notes in zeppelin:master/BaseLivyIntepreter.java" after inspecting source code and exploratory session with my laptop setup. If I opened two notes in zeppelin, LivyServer shows there is two session in their web UI. 

Did I missed something? Is it true `livy session is shared across notes`. Or Did @zjffdu wanted to know about rationale of PR. :)

**Comment by zjffdu** [comment]
@ruseel Livy session is shared across notes, as long as the notes belong to one user. Livy interpreter is designed for isolated across users. That means each user own one livy session, and he can share this session across his notes.  The multi-tenancy means users share the same hadoop cluster.

**Comment by zjffdu** [comment]
Right, one major reason is security. Sharing SparkContext across users may introduce potential security issue.

**Comment by felixcheung** [comment]
well, I think that is the "design" of the open source implementation ;)

**Comment by ruseel** [comment]
A goal of PR is sharing already created livy session, now I see 'there is a concern about security'.
so I am closing this PR.

**Pull request closed without merging**